src/index.js

```diff
@@ -0,0 +1,220 @@
+    this.compiledCircuitsPath = embark.config.dappPath('.embark', 'snarks');
+    this.registerEvents();
+  }
+
+  registerEvents() {
+    this.embark.registerActionForEvent(
+      // won't work as desired with "dev embark" (built in monorepo) until
+      // after first prerelease of embark v5
+      semver(this.embark.version).major >= 5 // || true // get rid of `|| true`
+        ? 'contracts:build:before'
+        : 'build:beforeAll',
+      (_, callback) => this.compileAndGenerateContracts(callback)
+    );
+  }
+
+  verifierFilepath(basename) {
+    return path.join(
+      this.compiledCircuitsPath,
+      `${basename}${Extension.VkVerifier}`
+    );
+  }
+
+  verifierContractPath(basename) {
+    return path.join(
+      this.compiledCircuitsPath,
+      `${basename}${Extension.Solidity}`
+    );
+  }
+
+  proofFilepath(basename) {
+    return path.join(
+      this.compiledCircuitsPath,
+      `${basename}${Extension.VkProof}`
+    );
+  }
+
+  async compileCircuits() {
+    await fs.ensureDir(this.compiledCircuitsPath);
+
+    const patterns = this.circuitsConfig.circuits;
+    if (!patterns) {
+      return;
+    }
+
+    this.embark.logger.info('Compiling circuits...');
+
+    await Promise.all(
+      patterns.map(async pattern => {
+        const filepaths = await glob(pattern);
+        await Promise.all(
+          filepaths
+            .filter(filename => path.extname(filename) === Extension.Circom)
+            .map(filepath => this.compileCircuit(filepath))
+        );
+      })
+    );
+  }
+
+  compileCircuit(filepath) {
+    return new Promise((resolve, reject) => {
+      const output = path.join(
+        this.compiledCircuitsPath,
+        `${path.basename(filepath, Extension.Circom)}${Extension.Json}`
+      );
+      exec(`${circomBinary} ${filepath} -o ${output}`, error => {
+        if (error) {
+          return reject(error);
+        }
+
+        resolve();
+      });
+    });
+  }
+
+  async generateProofs() {
+    this.embark.logger.info('Generating proofs...');
+
+    await Promise.all(
+      (await fs.readdir(this.compiledCircuitsPath))
+        .filter(filename => path.extname(filename) === Extension.Json)
+        .map(filename => this.generateProof(filename))
+    );
+  }
+
+  async generateProof(filename) {
+    const filepath = path.join(this.compiledCircuitsPath, filename);
+    const basename = path.basename(filename, Extension.Json);
+
+    const circuit = await this.getCircuit(filepath);
+    const setup = await this.generateSetup(circuit, basename);
+
+    const input = this.circuitsConfig.inputs[basename];
+    if (!input) {
+      return;
+    }
+
+    const witness = circuit.calculateWitness(input);
+    const { proof, publicSignals } = zkSnark.original.genProof(
+      setup.vk_proof,
+      witness
+    );
+    if (zkSnark.original.isValid(setup.vk_verifier, proof, publicSignals)) {
+      return this.generateVerifier(basename);
+    }
+
+    throw new Error(
```

Comment:
I'd personally put the `throw` in the `if` and reverse the condition. 
It's just easier to understand when the "normal" flow is outside of `if`s

Reply:
Done.